### PR TITLE
Make the atlas output `*.a.texturesetc` more deterministic

### LINKED ISSUE #9370: New hexDigest in atlas
Defold generates a new hexDigest for an atlas with multiple icons during each new build. This results in different checksums when in fact no changes have been made. Atlases with just one icon don't have this issue.

1. Create atlas with more than 1 images
2. Build with bob
3. Build again without any changes
4. Look at texturesetc hexDigest in game.graph of both builds

Expected behavior: hexDigest must be the same

Defold version: 1.6.2 (0dccda8a4b9d85d0e2f9a564e2cd326615028b4e)

Platform: Android, iOS

**Information from game.graph in the first build:**
{
  "path" : "/main/generated/_split_group_specific/23188_arcade/spine/A_decal_moon_03.a.texturesetc",
  "hexDigest" : "a9fec432ffc9a720bc653816ef7c61f3fafaeddd",
  "isInMainBundle" : false,
  "children" : [ "/main/generated/_split_group_specific/23188_arcade/spine/A_decal_moon_03.texturec" ]
}, {
  "path" : "/main/generated/_split_group_specific/23188_arcade/spine/A_decal_moon_03.texturec",
  "hexDigest" : "2f837b98c0dd212bdf6c2740d53b1e3af21b0f25",
  "isInMainBundle" : false,
  "children" : [ ]
}

**Information from game.graph in the second build:**
{
  "path" : "/main/generated/_split_group_specific/23188_arcade/spine/A_decal_moon_03.a.texturesetc",
  "hexDigest" : "20e7c16b765cecf8a1dffc81d5919c65e3fdbfe9",
  "isInMainBundle" : false,
  "children" : [ "/main/generated/_split_group_specific/23188_arcade/spine/A_decal_moon_03.texturec" ]
}, {
  "path" : "/main/generated/_split_group_specific/23188_arcade/spine/A_decal_moon_03.texturec",
  "hexDigest" : "2f837b98c0dd212bdf6c2740d53b1e3af21b0f25",
  "isInMainBundle" : false,
  "children" : [ ]
}

### PR DESCRIPTION
**Pull request by AGulev**
Fixes the issue where changing the order of images in an atlas without altering any content results in a different `*.a.texturesetc` file.

Fix https://github.com/defold/defold/issues/9370

## PR checklist

* [x] Code
	* [x] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [x] Documentation
	* [x] Make sure that API documentation is updated in code comments
	* [x] Make sure that manuals are updated (in github.com/defold/doc)
* [x] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [x] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [x] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/AbstractProtoBuilderTest.java [hunk 1]
```diff
@@ -175,6 +176,16 @@ protected List<Message> build(String file, String source) throws Exception {
...
 
+    protected void clean() throws Exception {
+        Collection<String> result = new ArrayList<>();
+        fileSystem.walk("build", new FileSystemWalker() {
+            public void handleFile(String path, Collection<String> results) {
+                fileSystem.addFile(path, null);
+            }
+        }, result);
+        project.build(new NullProgress(), "clean");
+    }
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/AbstractProtoBuilderTest.java:187]: It's important to clean up the result of a build if it's necessary to build the same file twice. For Mock FS, we can't delete files; we just replace their content with `nil`.

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/AtlasBuilderTest.java [hunk 2]
```diff
@@ -306,4 +310,45 @@ public void testAtlasFrameIndices() throws Exception {
         assertEquals(imageNameHashes, textureSet.getImageNameHashesList());
     }
 
+    @Test
+    public void testHexDigestForAtlas() throws Exception {
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/AtlasBuilderTest.java:314]: This is one of those tasks where it's easier to start with a test.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/AtlasUtil.java [hunk 3]
```diff
@@ -273,7 +286,7 @@ private static List<MappedAnimDesc> createAnimDescs(Atlas atlas, PathTransformer
 
             animDescs.add(new MappedAnimDesc(anim.getId(), framePaths, frameIds, anim.getPlayback(), anim.getFps(), anim.getFlipHorizontal() != 0, anim.getFlipVertical() != 0));
         }
-        for (AtlasImage image : atlas.getImagesList()) {
+        for (AtlasImage image : atlasImages) {
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/AtlasUtil.java:289]: Make sure that we use already sorted images.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/AtlasUtil.java [hunk 4]
```diff
@@ -168,6 +169,18 @@ public static List<AtlasImage> collectImages(IResource atlasResource, Atlas atla
                 }
             }
         }
+
+        images.sort(new Comparator<AtlasImage>() {
+            @Override
+            public int compare(AtlasImage img1, AtlasImage img2) {
+                int nameComparison = img1.getImage().compareTo(img2.getImage());
+                if (nameComparison == 0) {
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/AtlasUtil.java:177]: If name is the same, compare trim mode

**Review by JCash — APPROVED** [review]

**Pull request merged**